File: .py/61.RotateList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or k == 0:
            return head
        
        current = head
        length = 1
        while current.next:
            current = current.next
            length += 1
        current.next = head 
        
        k = k % length
        if k == 0:
            current.next = None 
            return head
        
        steps_to_new_head = length - k
        new_tail = head
        for _ in range(steps_to_new_head - 1):
            new_tail = new_tail.next
        
        new_head = new_tail.next
        new_tail.next = None  
        
        return new_head

File: .cpp/61.RotateList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *rotateRight(ListNode *head, int k)
    {
        if (!head || k == 0)
            return head;

        ListNode *current = head;
        int length = 1;
        while (current->next)
        {
            current = current->next;
            length += 1;
        }
        current->next = head;

        k = k % length;
        if (k == 0)
        {
            current->next = nullptr;
            return head;
        }

        int steps_to_new_head = length - k;
        ListNode *new_tail = head;
        for (int i = 0; i < steps_to_new_head - 1; ++i)
        {
            new_tail = new_tail->next;
        }

        ListNode *new_head = new_tail->next;
        new_tail->next = nullptr;

        return new_head;
    }
};